File: poh4/poh4_3.clj
#!/usr/bin/env clj
(set! *unchecked-math* true)
(defn run [a i cur r T N]
	(if (< i N)
		(let [val (read)]
			(aset-long ^longs a i val)
			(let [nxt (if (< i T) (+ cur val) (- (+ cur val) (aget ^longs a (- i T))))]
				(recur a (+ i 1) nxt (max r nxt) T N)
			)
		)
		r
	)
)
(let [T (read) N (read)]
	(def a (long-array N))
	(println (run a 0 0 0 T N))
)

File: poh4/poh4_3.hs
#!/usr/bin/env runghc
--import Debug.Trace
--debug code and seq (strict evaluation) from http://www.geocities.jp/m_hiroi/func/haskell05.html

import Control.Applicative
import Control.Monad

-- 1.82s/1.82s
run::[Int] -> [Int] -> Int -> Int -> Int
run a b r cur
	| null b = r
	| otherwise =
		let nxt = cur + (head b) - (head a) in
		nxt `seq` run (tail a) (tail b) (max r nxt) nxt
main = do
	[t,n] <- map (read :: String -> Int) . words <$> getLine
	a <- replicateM n (readLn :: IO Int)
	let s = sum (take t a)
	print $ run a (drop t a) s s

{-
-- 1.77s/1.77s
-- refer to https://gist.github.com/tanakh/7764127
import qualified Data.Vector.Unboxed as V
import Data.Vector.Unboxed ((!))
main = do
	[t,n] <- map (read :: String -> Int) . words <$> getLine
	_a <- replicateM n (readLn :: IO Int)
	let s = sum (drop (n-t) _a)
	let a = V.fromList _a
	let run i r cur
		| i==(-1) = r
		| otherwise =
			let nxt = cur + a!i - a!(i+t) in
			nxt `seq` run (i-1) (max r nxt) nxt
	print $ run (n-1-t) s s
-}

{-
mklist n
	| n>0 = do
		x <- (readLn :: IO Int)
		y <- mklist (n-1)
		return (x:y)
	| otherwise = return ([])
-}